WinRT.Runtime/AgileReference.cs

```diff
@@ -0,0 +1,97 @@
+﻿using System;
+using System.Runtime.InteropServices;
+using WinRT.Interop;
+
+namespace WinRT
+{
+    public class AgileReference : IDisposable
+    {
+        private readonly static Guid CLSID_StdGlobalInterfaceTable = Guid.Parse("00000323-0000-0000-c000-000000000046");
+        private readonly IAgileReference _agileReference;
+        private readonly IGlobalInterfaceTable _git;
+        private readonly IntPtr _cookie;
+        private bool disposed;
+
+        public unsafe AgileReference(IObjectReference instance)
+        {
+            IntPtr agileReference = default;
+            IntPtr gitPtr = default;
+            Guid iid = typeof(IUnknownVftbl).GUID;
+            try
+            {
+                Marshal.ThrowExceptionForHR(Platform.RoGetAgileReference(
+                    Platform.AgileReferenceOptions.AGILEREFERENCE_DEFAULT,
+                    ref iid,
+                    instance.ThisPtr,
+                    &agileReference));
+                _agileReference = ABI.WinRT.Interop.IAgileReference.FromAbi(agileReference).AsType<ABI.WinRT.Interop.IAgileReference>();
+            }
+            catch(TypeLoadException)
+            {
+                Guid gitClsid = CLSID_StdGlobalInterfaceTable;
+                Guid gitIid = typeof(IGlobalInterfaceTable).GUID;
+                Marshal.ThrowExceptionForHR(Platform.CoCreateInstance(
+                    ref gitClsid,
+                    IntPtr.Zero,
+                    1 /*CLSCTX_INPROC_SERVER*/,
+                    ref gitIid,
+                    &gitPtr));
+                _git = ABI.WinRT.Interop.IGlobalInterfaceTable.FromAbi(gitPtr).AsType<ABI.WinRT.Interop.IGlobalInterfaceTable>();
+                _cookie = _git.RegisterInterfaceInGlobal(instance, iid);
+
+            }
+            finally
+            {
+                MarshalInterface<IAgileReference>.DisposeAbi(agileReference);
+                MarshalInterface<IGlobalInterfaceTable>.DisposeAbi(gitPtr);
+            }
+        }
+
+        public IObjectReference Get() => _agileReference?.Resolve(typeof(IUnknownVftbl).GUID) ?? _git?.GetInterfaceFromGlobal(_cookie, typeof(IUnknownVftbl).GUID);
+
+        protected virtual void Dispose(bool disposing)
+        {
+            if (!disposed)
+            {
+                try
+                {
+                    if (_git != null && _cookie != IntPtr.Zero)
+                    {
+                        _git.RevokeInterfaceFromGlobal(_cookie);
+                    }
+                }
+                catch(ObjectDisposedException)
+                {
+                    // TODO: How to handle removing from git when it has already been disposed.

```

Comment:
were you actually seeing this? your dispose pattern is properly implemented.

Reply:
In the delegate scenarios I was seeing the finalizer already disposed those instance variables causing me to be unable to remove from the git.

Reply:
ok, I don't think it's an issue to leak a GIT cookie, but it so, we could add a wrapper at a later time (maybe change this comment to suggest that)

Reply:
> I don't think it's an issue to leak a GIT cookie

are you sure this won't pollute the GIT (i.e. actually leak memory) or cause access violations if someone tries to access the already-released COM pointer? Documentation says registered interfaces **_must_** stay alive until revoked.

Reply:
Ended up making the git table a static lazy reference which avoids this issue and avoids allocating it each time given it is a per process instance.